File: java/src/main/java/com/urlshortener/dal/repositories/shorturl/impl/ShortUrlRepositoryImpl.java
package com.urlshortener.dal.repositories.shorturl.impl;

import com.urlshortener.dal.entities.ShortUrl;
import com.urlshortener.dal.repositories.shorturl.ShortUrlRepository;
import com.urlshortener.dal.util.TypedQueryUtils;
import lombok.NonNull;

import javax.enterprise.context.RequestScoped;
import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import java.util.List;

@RequestScoped
public class ShortUrlRepositoryImpl implements ShortUrlRepository {

    @PersistenceContext
    private final EntityManager em;

    public ShortUrlRepositoryImpl(EntityManager entityManager) {this.em = entityManager;}

    @Override
    public ShortUrl merge(@NonNull ShortUrl shortUrl) {
        return this.em.merge(shortUrl);
    }

    @Override
    public void deleteById(long id, long userId) {
        var query = this.em.createQuery("delete from ShortUrl s where s.id = :id and s.user.id = :userId");
        query.setParameter("id", id);
        query.setParameter("userId", userId);
        query.executeUpdate();
    }

    @Override
    public Long countByShortName(@NonNull String shortName) {
        var query = this.em
                .createQuery("select count(s) from ShortUrl s where s.shortName = :shortName", Long.class);
        query.setParameter("shortName", shortName);
        return TypedQueryUtils.singleResult(query);
    }

    @Override
    public ShortUrl findByShortName(@NonNull String shortName) {
        var query = this.em
                .createQuery("select s from ShortUrl s where s.shortName = :shortName", ShortUrl.class);
        query.setParameter("shortName", shortName);
        return TypedQueryUtils.singleResult(query);
    }

    @Override
    public ShortUrl findByIdAndUserId(long id, long userId) {
        var query = this.em
                .createQuery("select s from ShortUrl s where s.id = :id and s.user.id = :userid", ShortUrl.class);
        query.setParameter("id", id);
        query.setParameter("userid", userId);
        return TypedQueryUtils.singleResult(query);
    }

    @Override
    public List<ShortUrl> findAllForUser(long userId) {
        var query =
                this.em.createQuery("select s from ShortUrl s where s.user.id = :userid", ShortUrl.class);
        query.setParameter("userid", userId);
        return query.getResultList();
    }
}

File: kotlin/src/main/kotlin/com/urlshortener/dal/repositories/shorturl/impl/ShortUrlRepositoryImpl.kt
@file:Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")

package com.urlshortener.dal.repositories.shorturl.impl

import com.urlshortener.dal.entities.ShortUrl
import com.urlshortener.dal.repositories.shorturl.ShortUrlRepository
import com.urlshortener.dal.util.singleResult
import javax.enterprise.context.RequestScoped
import javax.persistence.EntityManager
import javax.persistence.PersistenceContext
import java.lang.Long as JavaLong

@RequestScoped
class ShortUrlRepositoryImpl(@PersistenceContext private val em: EntityManager) : ShortUrlRepository {
    override fun merge(shortUrl: ShortUrl): ShortUrl = em.merge(shortUrl)

    override fun deleteById(id: Long, userId: Long) {
        em.createQuery("delete from ShortUrl s where s.id = :id and s.user.id = :userId").run {
            setParameter("id", id)
            setParameter("userId", userId)
            executeUpdate()
        }
    }

    override fun countByShortName(shortName: String): Long =
        em.createQuery("select count(s) from ShortUrl s where s.shortName = :shortName", JavaLong::class.java).run {
            setParameter("shortName", shortName)
            singleResult.toLong()
        }

    override fun findByShortName(shortName: String): ShortUrl? =
        em.createQuery("select s from ShortUrl s where s.shortName = :shortName", ShortUrl::class.java).run {
            setParameter("shortName", shortName)
            singleResult()
        }

    override fun findByIdAndUserId(id: Long, userId: Long): ShortUrl? =
        em.createQuery("select s from ShortUrl s where s.id = :id and s.user.id = :userid", ShortUrl::class.java).run {
            setParameter("id", id)
            setParameter("userid", userId)
            singleResult()
        }

    override fun findAllForUser(userId: Long): List<ShortUrl> =
        em.createQuery("select s from ShortUrl s where s.user.id = :userid", ShortUrl::class.java).run {
            setParameter("userid", userId)
            resultList
        }
}